lightning/src/ln/onion_utils.rs

```diff
@@ -1463,6 +1463,53 @@ where
 	}
 }
 
+/// Process failure we got back from upstream on a payment we sent (implying htlc_source is an
+/// OutboundRoute).
+pub fn process_onion_success<T: secp256k1::Signing, L: Deref>(
+	secp_ctx: &Secp256k1<T>, logger: &L, path: &Path, outer_session_priv: &SecretKey,
+	mut attribution_data: AttributionData,
+) -> Vec<u32>
+where
+	L::Target: Logger,
+{
+	let mut hold_times = Vec::new();
+
+	// Only consider hops in the regular path for attribution data. Blinded path attribution data isn't accessible.
+	let shared_secrets =
+		construct_onion_keys_generic(secp_ctx, &path.hops, None, outer_session_priv)
+			.map(|(shared_secret, _, _, _, _)| shared_secret);
+
+	// Path length can reach 27 hops, but attribution data can only be conveyed back to the sender from the first 20
+	// hops. Determine the number of hops to be used for attribution data.
+	let attributable_hop_count = usize::min(path.hops.len(), MAX_HOPS);
+
+	for (route_hop_idx, shared_secret) in shared_secrets.enumerate().take(attributable_hop_count) {
+		attribution_data.crypt(shared_secret.as_ref());
+
+		// Calculate position relative to the last attributable hop. The last attributable hop is at position 0. We need
+		// to look at the chain of HMACs that does include all data up to the last attributable hop. Hold times beyond
+		// the last attributable hop will not be available.
+		let position = attributable_hop_count - route_hop_idx - 1;
+		let hold_time = attribution_data.verify(&Vec::new(), shared_secret.as_ref(), position);
+		if let Some(hold_time) = hold_time {
+			hold_times.push(hold_time);
+
+			// Shift attribution data to prepare for processing the next hop.
+			attribution_data.shift_left();
+		} else {
+			log_debug!(
+				logger,
+				"Invalid HMAC in attribution data for node at pos {}",
+				route_hop_idx
+			);
+
+			break;
```

Comment:
It seems like we'll return a subset of the hold times in this case? Would it be better to `continue` and get the hold times that we can? I'm wondering if one node on the path can cause us to miss parsing some hold times here.